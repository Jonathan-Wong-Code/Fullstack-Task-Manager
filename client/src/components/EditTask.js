import React, { useEffect, useState } from "react";
import TaskForm from "./TaskForm";
import axios from "axios";
import { useTaskState } from "./../context/task-context";

export default function EditTask({ match }) {
  const [task, setTask] = useState();
  const { tasks } = useTaskState();

  useEffect(() => {
    const editedTask = tasks.find(task => task._id === match.params.id);
    const getTask = async () => {
      const response = await axios({
        method: "GET",
        url: `http://localhost:3000/api/v1/tasks/${match.params.id}`,
        withCredentials: true
      });
      setTask(response.data.data.task);
    };
    if (editedTask) {
      setTask(editedTask);
    } else {
      getTask();
    }
  }, [tasks, match.params.id]);

  if (!task) return <div />;
  return (
    <div>
      <h2>Edit Task</h2>
      <TaskForm type="edit" editedTask={task} />
    </div>
  );
}
