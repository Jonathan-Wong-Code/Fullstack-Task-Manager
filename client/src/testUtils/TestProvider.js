import React from "react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";

function TestProvider({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <TaskProvider>{children}</TaskProvider>
      </UserProvider>
    </AuthProvider>
  );
}

export default TestProvider;
