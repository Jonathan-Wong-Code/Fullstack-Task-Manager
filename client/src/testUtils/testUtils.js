import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render } from "@testing-library/react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";
import React from "react";
export const renderRouter = (
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    ...options
  }
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>, options),
    history
  };
};

export function renderWithContextRouter(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    authValue,
    userValue,
    taskValue,
    ...options
  }
) {
  return {
    ...render(
      <AuthProvider value={authValue}>
        <UserProvider value={userValue}>
          <TaskProvider value={taskValue}>
            <Router history={history}>{ui}</Router>
          </TaskProvider>
        </UserProvider>
      </AuthProvider>,
      options
    ),
    history
  };
}
