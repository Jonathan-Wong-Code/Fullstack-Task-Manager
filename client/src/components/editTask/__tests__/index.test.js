import React from "react";
import EditTask from "..";
import { cleanup, wait } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { renderWithContextRouter } from "./../../../testUtils/testUtils";
import axios from "axios";

jest.mock("axios");

afterEach(cleanup);

const mockData = {
  data: {
    data: {
      task: {
        completed: true,
        title: "task 1",
        description: "task 1"
      }
    }
  }
};

describe("<EditTask />", () => {
  axios.get.mockImplementation(() => Promise.resolve(mockData));

  const match = { params: { id: "test123" } };

  test("it renders with pre-existing task", async () => {
    const testUser = { name: "Jon", email: "jon@jon.com" };
    const { getByLabelText } = renderWithContextRouter(
      <EditTask match={match} />,
      {
        route: "/create",
        authValue: testUser
      }
    );

    //Integration with TaskForm
    await wait(() => {
      const title = getByLabelText("Title:");
      const description = getByLabelText("Description:");
      const completed = getByLabelText("completed:");
      expect(title).toHaveValue("task 1");
      expect(description).toHaveValue("task 1");
      expect(completed).toHaveAttribute("value", "true");
    });
  });
});
