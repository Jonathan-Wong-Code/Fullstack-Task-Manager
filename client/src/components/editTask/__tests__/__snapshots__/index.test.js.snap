// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditTask /> it renders 1`] = `
<div>
  <div>
    <h2>
      Edit Task
    </h2>
    <form>
      <label
        for="title"
      >
        Title:
      </label>
      <input
        id="title"
        type="text"
        value="task 1"
      />
      <label
        for="description"
      >
        Description:
      </label>
      <textarea
        cols="50"
        id="description"
        rows="10"
        type="text"
      >
        task 1
      </textarea>
      <label
        for="completed"
      >
        completed?
      </label>
      <input
        checked=""
        id="completed"
        type="checkbox"
        value="true"
      />
      <button>
         
        Edit
      </button>
    </form>
    
    
  </div>
</div>
`;
